trl/trainer/orpo_trainer.py

```diff
@@ -756,51 +767,79 @@ def concatenated_forward(
             concatenated_batch["concatenated_input_ids"],
```

Comment:
I don't think this will get the full benefits of the liger kernel. This will perform full model forward pass including the last layer, and will materialize the full logits: https://github.com/huggingface/transformers/blob/eed11f34abe558c76590366a21316318e5f820ed/src/transformers/models/llama/modeling_llama.py#L1179 

Using `output_hidden_states` and then using Liger's loss based on the last hidden state + lm head is not enough. We need to do the forward pass only until the lm head. 

For Llama, we can do the forward pass on just the `LlamaModel` submodule (`model.model`) : 

```
# tiny random llama repr
LlamaForCausalLM(
  (model): LlamaModel(
    (embed_tokens): Embedding(32000, 16, padding_idx=31999)
    (layers): ModuleList(
      (0-1): 2 x LlamaDecoderLayer(
        (self_attn): LlamaSdpaAttention(
          (q_proj): Linear(in_features=16, out_features=16, bias=False)
          (k_proj): Linear(in_features=16, out_features=16, bias=False)
          (v_proj): Linear(in_features=16, out_features=16, bias=False)
          (o_proj): Linear(in_features=16, out_features=16, bias=False)
          (rotary_emb): LlamaRotaryEmbedding()
        )
        (mlp): LlamaMLP(
          (gate_proj): Linear(in_features=16, out_features=64, bias=False)
          (up_proj): Linear(in_features=16, out_features=64, bias=False)
          (down_proj): Linear(in_features=64, out_features=16, bias=False)
          (act_fn): SiLU()
        )
        (input_layernorm): LlamaRMSNorm((16,), eps=1e-06)
        (post_attention_layernorm): LlamaRMSNorm((16,), eps=1e-06)
      )
    )
    (norm): LlamaRMSNorm((16,), eps=1e-06)
    (rotary_emb): LlamaRotaryEmbedding()
  )
  (lm_head): Linear(in_features=16, out_features=32000, bias=False)
)
```

I believe the implementation might end up being model specific because we have to get the attribute name for this common backbone first (I'm not sure if all HF models have the backbone under ".model" attribute). For example, see impl here https://github.com/linkedin/Liger-Kernel/blob/0bb6c72d03f0fc570b9e954c42b723ab7e0c315f/src/liger_kernel/transformers/trainer/orpo_trainer.py#L93-L110

Reply:
thanks @SumanthRH can you check the current implementation of how i was thinking about skipping the lm_head layer?